### PR TITLE
build: Moving more downstream code into the listener extension

### PR DESCRIPTION
**Pull request by alyssawilk**
Risk Level: low
Testing: n/a
Docs Changes: n/a
Release Notes: n/a

part of https://github.com/envoyproxy/envoy-mobile/issues/2711

### CODE REVIEW

File: source/server/server.cc [hunk 1]
```diff
@@ -64,6 +63,19 @@
...
 namespace Server {
+namespace {
+std::unique_ptr<ConnectionHandler> getHandler(Event::Dispatcher& dispatcher) {
+
+  auto* factory = Config::Utility::getFactoryByName<ConnectionHandlerFactory>(
+      "envoy.connection_handler.default");
+  if (factory) {
+    return factory->createConnectionHandler(dispatcher, absl::nullopt);
+  }
+  ENVOY_LOG_MISC(debug, "Unable to find envoy.connection_handler.default factory");
```
Comment (soulxu) [inline source/server/server.cc:74]: Why not raise an exception? if returns nullptr, it will crash Envoy.

Comment (alyssawilk) [reply source/server/server.cc:74]: this should never happen in upstream Envoy.  it can in Envoy mobile if you misconfigure things.

Comment (snowp) [reply source/server/server.cc:74]: Would that be due to failing to register the factory? Or is there some config that drives this?

In general EM does crash on missing registration, though I guess just logging might be better?

**Review by snowp — APPROVED** [review]
Code move lgtm

File: source/server/server.cc (hunk 1 above)
Comment (Augustyniak) [reply source/server/server.cc:74]: This is hit in Envoy Mobile which makes mobile application crash after we call `addDirectResponse` on `TestEngineBuilder` (as it adds a listener) (iOS only API).

**Pull request merged**